core/sdk-core/src/main/java/software/amazon/awssdk/core/progress/snapshot/DefaultProgressSnapshot.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.core.progress.snapshot;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Optional;
+import java.util.OptionalDouble;
+import java.util.OptionalLong;
+import java.util.concurrent.TimeUnit;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.utils.Validate;
+import software.amazon.awssdk.utils.builder.CopyableBuilder;
+import software.amazon.awssdk.utils.builder.ToCopyableBuilder;
+
+/**
+ * An SDK-internal implementation of {@link ProgressSnapshot}.
+ */
+@SdkInternalApi
+public class DefaultProgressSnapshot
+    implements ToCopyableBuilder<DefaultProgressSnapshot.Builder, DefaultProgressSnapshot>,
+               ProgressSnapshot  {
+
+    private final long transferredBytes;
+    private final Long totalBytes;
+    private final Optional<Instant> startTime;
+
+    public DefaultProgressSnapshot(Builder builder) {
+        if (builder.totalBytes != null) {
+            Validate.isNotNegative(builder.totalBytes, "totalBytes");
+            Validate.isTrue(builder.transferredBytes <= builder.totalBytes,
+                            "transferredBytes (%s) must not be greater than totalBytes (%s)",
+                            builder.transferredBytes, builder.totalBytes);
+        }
+        Validate.paramNotNull(builder.transferredBytes, "byteTransferred");
+        this.transferredBytes = Validate.isNotNegative(builder.transferredBytes, "transferredBytes");
+        this.totalBytes = builder.totalBytes;
+
+        if (builder.startTime.isPresent()) {
+            Instant currentTime = Instant.now();
```

Comment:
This is an internal class that is only used by us. This validation and thus extra compute time doesn't seem to be necessary IMO.

Reply:
Fair point. We should always have a start time for this default implementation class.